src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,245 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import pinkie from "pinkie";
+import {
+  merge as observableMerge,
+  Observable,
+  ReplaySubject,
+} from "rxjs";
+import {
+  catchError,
+  filter,
+  map,
+  mergeMap,
+  take,
+} from "rxjs/operators";
+import createSegmentLoader from "../../../core/pipelines/segment/create_segment_loader";
+import Player from "../../../index";
+import { ISegment } from "../../../manifest";
+import dash from "../../../transports/dash";
+import getContentInfos from "./get_content_infos";
+import {
+  disposeSourceBuffer,
+  initSourceBuffer$,
+} from "./init_source_buffer";
+import log from "./log";
+import removeBufferForTime$ from "./remove_buffer";
+import {
+  getNeededSegment,
+  IContentInfos,
+} from "./utils";
+import VideoThumbnailLoaderError from "./video_thumbnail_loader_error";
+
+const PPromise = typeof Promise === "function" ? Promise :
+                                                 pinkie;
+
+interface IJob { contentInfos: IContentInfos;
+                 segment: ISegment;
+                 stop: () => void;
+                 jobPromise: Promise<unknown>; }
+
+const segmentLoader = createSegmentLoader(
+  dash({ lowLatencyMode: false }).video.loader,
+  { maxRetry: 0,
+    maxRetryOffline: 0,
+    initialBackoffDelay: 0,
+    maximumBackoffDelay: 0, }
+);
+
+/**
+ * This tool, as a supplement to the RxPlayer, intent to help creating thumbnails
+ * from a video source.
+ *
+ * The tools will extract a "thumbnail track" either from a video track (whose light
+ * chunks are adapted from such use case) or direclty from the media content.
+ */
+export default class VideoThumbnailLoader {
+  private readonly _videoElement: HTMLVideoElement;
+
+  private _player: Player;
+  private _currentJob?: IJob;
+
+  constructor(videoElement: HTMLVideoElement,
+              player: Player) {
+    this._videoElement = videoElement;
+    this._player = player;
+  }
+
+  /**
+   * Set time of thumbnail video media element :
+   * - Remove buffer when too much buffered data
+   * - Search for thumbnail track element to display
+   * - Load data
+   * - Append data
+   * Resolves when time is set.
+   * @param {number} time
+   * @returns {Promise}
+   */
+  setTime(time: number): Promise<unknown> {
+    return this._setTime(time, this._videoElement);
+  }
+
+  /**
+   * Dispose thumbnail loader.
+   * @returns {void}
+   */
+  dispose(): void {
+    disposeSourceBuffer(this._videoElement);
+    this._currentJob?.stop();
+  }
+
+  private _setTime = (time: number,
+                      videoElement: HTMLVideoElement
```

Comment:
I don't get why creating a new function is needed here

Reply:
Legacy. I remove it